source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc

```diff
@@ -72,6 +78,24 @@ DnsCacheImpl::loadDnsCacheEntry(absl::string_view host, uint16_t default_port,
   }
 }
 
+ResourceLimit& DnsCacheImpl::onDnsRequest(const Router::RouteEntry* route_entry,
+                                          Upstream::ClusterInfoConstSharedPtr cluster_info) {
```

Comment:
A few things:
1) Call this method `canCreateDnsRequest`
2) Don't pass RouteEntry, as we may use this outside of route context. Just pass the priority.
3) In the case of overflow inside this function, I would return an optional reference to a ResourceLimit, then you don't have to call `canCreate()` in the calling function.

Reply:
I would make this method signature as `ResourceAutoIncDecPtr DnsCacheImpl::canCreateDnsRequest(absl::optional<std::reference_wrapper<ResourceLimit>> cluster_resource)`, so:
- not leaking RouteEntry/ClusterInfo
- not calling canCreate twice.

WDYT?